enhancements/rhcos/rhcos-inject.md

```diff
@@ -0,0 +1,198 @@
+
+## Motivation
+
+Admins need a way to inject customization into their RHCOS nodes before they are provisioned by the cluster. In most cases, this configuration is require in order for the provisioning process to complete, so the normal facilities (e.g. Machine Configuration Operator) are not yet available. Today, this is a very manual process involving an admin interactively providing that configuration at the point of running the RHCOS installer. This might work in some environments, but many others don't have easy interactive access due to technical or policy constraints.
+
+### Goals
+
+- Make it easy for an admin to add network configuration to each of the RHCOS nodes added both before and after OpenShift installation.
+- Allow an admin to add custom services and miscellaneous configuration to their RHCOS nodes.
+
+### Non-Goals
+
+- Inventory management
+- Bring-your-own-RHEL customization
+
+## Proposal
+
+Add a new, optional step to the installation process that allows an admin to inject network configuration and any other customization to the RHCOS installer image (and maybe other RHCOS images in the future) before invocation. This customization is performed by a new `rhcos-inject` utility that takes as input an RHCOS installer image (`rhcos-<version>-installer.<architecture>.iso`), network configuration, and a set of Ignition Configs and generates a new RHCOS installer image with all of the configuration and customization included. This new installer image can then be booted in an unsupervised manner and will complete the RHCOS installation along with any customization.
+
+Initially, `rhcos-inject` will primarily focus on network configuration but as the needs of customers evolve, this can be expanded to include other common customization. As always, an escape hatch is needed for unanticipated cases and this will be in the form of raw Ignition Configs. If there is a need for additional customization beyond network configuration, the admin can include that in an Ignition Config and inject that alongside everything else.
+
+An example of its invocation can be seen here:
+
+```console
+$ rhcos-inject \
+	--input=rhcos-installer.x86_64.iso \
+	--output=control-0.iso \
+	--openshift-config=master.ign \
+	--bond=bond0:em1,em2:mode=active-backup \
+	--ip=10.10.10.2::10.10.10.254:255.255.255.0:control0.example.com:bond0:none
+```
+
+With a little scripting, a number of custom RHCOS installers can be quickly created:
+
+```zsh
+for i in {00..11}
+do
+	rhcos-inject \
+		--input=rhcos-installer.x86_64.iso \
+		--output=worker-${i}.iso \
+		--openshift-config=worker.ign \
+		--ip=10.10.10.$((i+10))::10.10.10.254:255.255.255.0:worker${i}.example.com:enp1s0:none
+done
+```
```

Comment:
This is basically a workaround for the fact that MCO can't do per-machine configs ref https://github.com/openshift/machine-config-operator/issues/1720 right?

This won't work for IPI baremetal as it's currently implemented since we don't have a way to customize the ignition files per host, we'd have to implement some equivalent customization method in the deploy workflow.

Also I wonder what do we do on day-2 - to enable the configuration to me managed we'd need to also provide equivalent config via MachineConfig, but there's currently no way to do that per-machine - what happens for example if some syntax change is required on upgrade to the nic configs that result from this process?